heat/core/dndarray.py

```diff
@@ -248,6 +259,111 @@ def strides(self):
+        except IndexError:
+            print("Indices out of bound")
+
+        return self.__array[ix].clone().contiguous()
+
+    def gethalo(self, halo_size):
+        """
+        Fetch halos of size 'halo_size' from neighboring ranks and save them in self.halo_next/self.halo_prev
+        in case they are not already stored. If 'halo_size' differs from the size of already stored halos,
+        the are overwritten.
+
+        Parameters
+        ----------
+        halo_size : int
+            Size of the halo. If halo_size exceeds the size of the HeAT tensor in self.split direction
+            the whole local tensor.array will be fetched
+
+        Returns
+        -------
+        None
+        """
+
+        if not isinstance(halo_size, int):
+            raise TypeError(
+                "halo_size needs to be of Python type integer, {} given)".format(type(halo_size))
+            )
+        if halo_size < 0:
+            raise ValueError(
+                "halo_size needs to be a positive Python integer, {} given)".format(type(halo_size))
+            )
+
+        if self.comm.is_distributed() and self.split is not None:
+            min_chunksize = self.shape[self.split] // self.comm.size
+            if halo_size > min_chunksize:
+                raise ValueError(
+                    "halo_size {} needs to smaller than chunck-size {} )".format(
+                        halo_size, min_chunksize
+                    )
+                )
+
+        if self.comm.is_distributed() and halo_size > 0 and self.split is not None:
+
+            a_prev = self.__prephalo(0, halo_size)
+            a_next = self.__prephalo(-halo_size, None)
+
+            res_prev = None
+            res_next = None
+
+            if self.comm.rank != self.comm.size - 1:
+                self.comm.Isend(a_next, self.comm.rank + 1)
+                res_prev = torch.zeros(a_prev.size(), dtype=a_prev.dtype)
+                req = self.comm.Irecv(res_prev, source=self.comm.rank + 1)
+                req.Wait()
+
+            if self.comm.rank != 0:
+                self.comm.Isend(a_prev, self.comm.rank - 1)
+                res_next = torch.zeros(a_next.size(), dtype=a_next.dtype)
+                req = self.comm.Irecv(res_next, source=self.comm.rank - 1)
+                req.Wait()
+
+            self.__halo_next = res_prev
+            self.__halo_prev = res_next
+            self.__ishalo = True
+
+    def cathalo(self):
+        """
+        Fetch halos of size 'halo_size' from neighboring ranks and save them in self.halo_next/self.halo_prev
+        in case they are not already stored. If 'halo_size' differs from the size of already stored halos,
+        the are overwritten.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        array + halos: pytorch tensors
+        """
+        return torch.cat(
+            tuple(_ for _ in (self.__halo_prev, self.__array, self.__halo_next) if _ is not None),
```

Comment:
possible to do a list comprehension here instead of tuple? would avoid one copy

Reply:
fixed